rpc/lib/client/ws_client.go

```diff
@@ -1,160 +1,390 @@
+			return errors.Wrap(err, "reached maximum reconnect attempts")
+		}
+	}
+}
+
+func (c *WSClient) startReadWriteRoutines() {
+	c.wg.Add(2)
+	c.receiveRoutineQuit = make(chan struct{})
+	go c.receiveRoutine()
+	go c.writeRoutine()
+}
+
+func (c *WSClient) reconnectRoutine() {
+	for {
+		select {
+		case originalError := <-c.reconnectAfter:
+			// wait until writeRoutine and receiveRoutine finish
+			c.wg.Wait()
+			err := c.reconnect()
+			if err != nil {
+				c.Logger.Error("failed to reconnect", "err", err, "original_err", originalError)
+				c.Stop()
+				return
+			} else {
+				// drain reconnectAfter
+			LOOP:
+				for {
+					select {
+					case <-c.reconnectAfter:
+					default:
+						break LOOP
+					}
+				}
+				c.startReadWriteRoutines()
+			}
+		case <-c.Quit:
+			return
+		}
+	}
+}
+
+// The client ensures that there is at most one writer to a connection by
+// executing all writes from this goroutine.
+func (c *WSClient) writeRoutine() {
+	ticker := time.NewTicker(c.pingPeriod)
+	defer func() {
+		ticker.Stop()
+		c.conn.Close()
+		c.wg.Done()
+	}()
+
+	for {
+		select {
+		case request := <-c.backlog:
+			c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			err := c.conn.WriteJSON(request)
 			if err != nil {
-				wsc.Logger.Info("WSClient failed to parse message", "err", err, "data", string(data))
-				wsc.ErrorsCh <- err
-				continue
+				c.Logger.Error("failed to resend request", "err", err)
+				c.reconnectAfter <- err
+				// add request to the backlog, so we don't lose it
+				c.backlog <- request
+				return
 			}
-			if response.Error != "" {
-				wsc.ErrorsCh <- errors.Errorf(response.Error)
-				continue
+			c.Logger.Info("resend a request", "req", request)
+		case request := <-c.send:
+			c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			err := c.conn.WriteJSON(request)
+			if err != nil {
+				c.Logger.Error("failed to send request", "err", err)
+				c.reconnectAfter <- err
+				// add request to the backlog, so we don't lose it
+				c.backlog <- request
+				return
 			}
-			wsc.ResultsCh <- *response.Result
+		case <-ticker.C:
+			c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			err := c.conn.WriteMessage(websocket.PingMessage, []byte{})
+			if err != nil {
+				c.Logger.Error("failed to write ping", "err", err)
+				c.reconnectAfter <- err
+				return
+			}
+			c.mtx.Lock()
+			c.sentLastPingAt = time.Now()
+			c.mtx.Unlock()
+			c.Logger.Debug("sent ping")
+		case <-c.receiveRoutineQuit:
+			return
+		case <-c.Quit:
+			c.conn.WriteMessage(websocket.CloseMessage, []byte{})
+			return
 		}
 	}
-	// this must be modified in the same go-routine that reads from the
-	// connection to avoid race conditions
-	wsc.Conn = nil
+}
+
+// The client ensures that there is at most one reader to a connection by
+// executing all reads from this goroutine.
+func (c *WSClient) receiveRoutine() {
+	defer func() {
+		c.conn.Close()
+		c.wg.Done()
+	}()
+
+	c.conn.SetReadDeadline(time.Now().Add(c.pongWait))
```

Comment:
-> https://github.com/tendermint/tendermint/pull/604#discussion_r131510952

no, we need this deadline